models/software_version_helpers.go

```diff
@@ -0,0 +1,199 @@
+func (u *UpdateServiceSoftwareVersionsParams) Validate() error {
+	if u.SoftwareVersions == nil {
+		return nil
+	}
+
+	for _, sv := range *u.SoftwareVersions {
+		switch sv.Name {
+		case MysqldSoftwareName:
+		case XtrabackupSoftwareName:
+		case XbcloudSoftwareName:
+		case QpressSoftwareName:
+		default:
+			return errors.Wrapf(ErrInvalidArgument, "invalid software name %q", sv.Name)
+		}
+	}
+
+	return nil
+}
+
+// UpdateServiceSoftwareVersions updates existing service software versions.
+func UpdateServiceSoftwareVersions(
+	q *reform.Querier,
+	serviceID string,
+	params UpdateServiceSoftwareVersionsParams,
+) (*ServiceSoftwareVersions, error) {
+	if err := params.Validate(); err != nil {
+		return nil, err
+	}
+
+	row, err := FindServiceSoftwareVersions(q, serviceID)
+	if err != nil {
+		return nil, err
+	}
+	if params.NextCheckAt != nil {
+		row.NextCheckAt = *params.NextCheckAt
+	}
+	if params.SoftwareVersions != nil {
+		row.SoftwareVersions = *params.SoftwareVersions
+	}
+
+	if err := q.Update(row); err != nil {
+		return nil, errors.Wrap(err, "failed to update service software versions")
+	}
+
+	return row, nil
+}
+
+// FindServiceSoftwareVersions returns service software versions entry by given service ID if found, ErrNotFound if not.
+func FindServiceSoftwareVersions(q *reform.Querier, serviceID string) (*ServiceSoftwareVersions, error) {
+	if serviceID == "" {
+		return nil, errors.New("service id is empty")
+	}
+
+	versions := &ServiceSoftwareVersions{ServiceID: serviceID}
+	switch err := q.Reload(versions); err {
+	case nil:
+		return versions, nil
+	case reform.ErrNoRows:
+		return nil, errors.Wrapf(ErrNotFound, "service software versions by service id '%s'", serviceID)
+	default:
+		return nil, errors.WithStack(err)
+	}
+}
+
+// FindServicesSoftwareVersionsFilter represents a filter for finding service software versions.
+type FindServicesSoftwareVersionsFilter struct {
+	Limit *int
+}
+
+// FindServicesSoftwareVersions returns all services software versions.
+func FindServicesSoftwareVersions(q *reform.Querier, filter FindServicesSoftwareVersionsFilter) ([]*ServiceSoftwareVersions, error) {
+	var args []interface{}
+	var tail strings.Builder
+	tail.WriteString("ORDER BY next_check_at ")
+	if filter.Limit != nil {
+		tail.WriteString("LIMIT $1")
+		args = append(args, *filter.Limit)
+	}
+
+	structs, err := q.SelectAllFrom(ServiceSoftwareVersionsTable, tail.String(), args...)
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+
+	versions := make([]*ServiceSoftwareVersions, len(structs))
+	for i, s := range structs {
+		versions[i] = s.(*ServiceSoftwareVersions)
+	}
+
+	return versions, nil
+}
+
+// DeleteServiceSoftwareVersions removes entry from the DB by service ID.
+func DeleteServiceSoftwareVersions(q *reform.Querier, serviceID string) error {
+	if _, err := FindServiceSoftwareVersions(q, serviceID); err != nil {
+		return err
+	}
+
+	if err := q.Delete(&ServiceSoftwareVersions{ServiceID: serviceID}); err != nil {
```

Comment:
this method deletes only one record by primary key, so probably it won't work.

Reply:
but why? I have a test for this method :)

Reply:
Oh, I see, we have one record for each service.